modin/experimental/cloud/cluster.py

```diff
@@ -0,0 +1,188 @@
+
+import os
+import errno
+from typing import NamedTuple
+import atexit
+
+from modin import set_backends
+
+from .base import ConnectionDetails
+from .connection import Connection
+
+
+class _RegionZone(NamedTuple):
+    region: str
+    zone: str
+
+
+class Provider:
+    AWS = "aws"
+
+    __KNOWN = {AWS: [_RegionZone(region="us-west-1", zone="us-west-1a")]}
+    __DEFAULT_HEAD = {AWS: "m5.large"}
+    __DEFAULT_WORKER = {AWS: "m5.large"}
+
+    def __init__(
+        self,
+        name: str,
+        credentials_file: str = None,
+        region: str = None,
+        zone: str = None,
+    ):
+        """
+        Class that holds all information about particular connection to cluster provider, namely
+            * provider name (must be one of known ones)
+            * path to file with credentials (file format is provider-specific); omit to use global provider-default credentials
+            * region and zone where cluster is to be spawned (optional, would be deduced if omitted)
+        """
+
+        if name not in self.__KNOWN:
+            raise ValueError(f"Unknown provider name: {name}")
+        if credentials_file is not None and not os.path.exists(credentials_file):
+            raise OSError(
+                errno.ENOENT, "Credentials file does not exist", credentials_file
+            )
+
+        if region is None:
+            if zone is not None:
+                raise ValueError("Cannot specify a zone without specifying a region")
+            try:
+                region, zone = self.__KNOWN[name][0]
+            except IndexError:
+                raise ValueError(f"No defaults for provider {name}")
+        elif zone is None:
+            for regzone in self.__KNOWN[name]:
+                if regzone.region == region:
+                    zone = regzone.zone
+                    break
+            else:
+                raise ValueError(f"No default for region {region} for provider {name}")
+
+        self.name = name
+        self.region = region
+        self.zone = zone
+        self.credentials_file = (
+            os.path.abspath(credentials_file) if credentials_file is not None else None
+        )
+
+    @property
+    def default_head_type(self):
+        return self.__DEFAULT_HEAD[self.name]
+
+    @property
+    def default_worker_type(self):
+        return self.__DEFAULT_WORKER[self.name]
+
+
+class BaseCluster:
+    """
+    Cluster manager for Modin. Knows how to use certain tools to spawn and destroy clusters,
+    can serve as context manager to switch execution engine and partition to remote.
+    """
+
+    target_engine = None
+    target_partition = None
+
+    def __init__(
+        self,
+        provider: Provider,
+        project_name: str = None,
+        cluster_name: str = "modin-cluster",
+        worker_count: int = 4,
+        head_node_type: str = None,
+        worker_node_type: str = None,
+    ):
+        """
+        Prepare the cluster manager. It needs to know a few things:
```

Comment:
Please add more detailed parameter documentation in the docstring, since it will be user facing. This docstring will be used to build our documentation on the API.

Reply:
I'm going to leave the docstring for _this_ class as is, because this class is not expected to be used directly, but rather via `cluster.create()` API, where I am adding the detailed docstring.